datafusion/functions/src/datetime/date_trunc.rs

```diff
@@ -47,6 +47,71 @@ use chrono::{
     DateTime, Datelike, Duration, LocalResult, NaiveDateTime, Offset, TimeDelta, Timelike,
 };
 
+/// Represents the granularity for date truncation operations
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum DateTruncGranularity {
+    Microsecond,
+    Millisecond,
+    Second,
+    Minute,
+    Hour,
+    Day,
+    Week,
+    Month,
+    Quarter,
+    Year,
+}
+
+impl DateTruncGranularity {
+    /// Mapping of string representations to enum variants
+    const GRANULARITY_MAP: &[(&str, Self)] = &[
+        ("microsecond", Self::Microsecond),
+        ("millisecond", Self::Millisecond),
+        ("second", Self::Second),
+        ("minute", Self::Minute),
+        ("hour", Self::Hour),
+        ("day", Self::Day),
+        ("week", Self::Week),
+        ("month", Self::Month),
+        ("quarter", Self::Quarter),
+        ("year", Self::Year),
+    ];
+
+    /// Parse a granularity string into a DateTruncGranularity enum
+    fn from_str(s: &str) -> Result<Self> {
+        let s_lower = s.to_lowercase();
+        Self::GRANULARITY_MAP
+            .iter()
+            .find(|(key, _)| *key == s_lower.as_str())
```

Comment:
Perhaps we can make this faster via https://doc.rust-lang.org/std/vec/struct.Vec.html#method.binary_search?

Or maybe just hard coding the strings so the compiler can make a jump table

Reply:
yeah, Rust still doesn't support const enums :(